Add typescript interface definition

sedenardi opened the pull request:
This PR satisfies https://github.com/cornerstonejs/dicomParser/issues/124 by adding the `index.d.ts` definition file to the project root. This is all that's required for Typescript projects to pick up the definitions.

I've copied the definitions from the file suggested in https://github.com/cornerstonejs/dicomParser/issues/124, as well as added the definitions to the `parseDicom` options.

Let me know if you have any questions/comments/concerns. Thank you for your work on this amazing library.

sedenardi commented:
Sorry for the last minute addition, noticed I hadn't brought the `Element` definition into the branch when I made the PR.

malaterre commented:
I am surprised by this in class `Element`:

`length?: number;`

I would I thought this would be:

`length: number;`

malaterre commented:
Another thing, where is `parser`:

* https://github.com/cornerstonejs/dicomParser/blob/master/src/dataSet.js#L25

sedenardi commented:
> I am surprised by this in class Element:
> length?: number;
> I would I thought this would be:
> length: number;
> Same goes for dataOffset I do not believe it can be undefined

These were oversights on my part. You're right, in both `readDicomElementExplicit` and `readDicomElementImplicit` the `length` and `dataOffset` are always set.

> Another thing, where is parser

I've added the optional `parser` member to the Element interface.

I've also added annotations to the `DataSet` data typed methods to show their return type and which VRs they're applicable to.

pavelkeyzik commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * Finds the element for tag and returns a string if it exists and has data. Use this function for VR types AE, CS, SH, and LO.
+     */
+    string: (tag: string, index: number) => string;
```
Sorry guys, can you explain to me why the index parameter is required? Works pretty well without it.

sedenardi replied:
Good catch @pavelkeyzik, index should be optional.

I did not make a default parameter in the interface since the implementation doesn't use a default parameter, but rather assigns it to `0` in the function's body.

dannyrb commented:
If @malaterre, is happy, it has my greenlight for merge

malaterre commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    byteArray: ByteArray;
+    byteArrayParser : ByteArrayParser;
+    elements: {
+      [key: string]: Element;
```
Replace `key: string` with `tag: string`, since this is what you used in your other function. Maybe even add a comment saying: `Access Element with tag: xGGGGEEEE`.

sedenardi replied:
Good suggestion. Added.

malaterre commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    elements: {
+      [key: string]: Element;
+    };
+    warnings: string[];
```
I believe `warnings` are optional, right ?

sedenardi replied:
I think it's always present. The `DataSet` gets its `warning` field set from the `ByteStream` [after it's instantiated](https://github.com/cornerstonejs/dicomParser/blob/374ad3d20a92bd2a7ec56d90dfce2fb2815fee13/src/parseDicom.js#L124), and `ByteStream` [always has](https://github.com/cornerstonejs/dicomParser/blob/374ad3d20a92bd2a7ec56d90dfce2fb2815fee13/src/byteStream.js#L45) a `warnings` value.

Please let me know if I'm missing something though.

malaterre commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    uint32: (tag: string, index?: number) => number;
+
+    /**
+     * Finds the element for tag and returns an signed int 32 if it exists and has data. Use this function for VR type SL.
```
`an signed` => `a signed`

sedenardi replied:
Fixed typos.

malaterre commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    double: (tag: string, index?: number) => number;
+
+    /**
+     * Returns the Value Multiplicity of an elements - the number of values in a multi-valued element.
```
`the Value Multiplicity` should read: `the actual Value Multiplicity` to be clear with regards to the DICOM standard. I am not sure what this function does in the case of VR=UN. Also `an elements` => `an element`

sedenardi replied:
Changed the description and fixed the typo.

For VR=UN, wouldn't the function still return the actual number of values separated by the backslash delimiter, based on its string representation?

malaterre commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    attributeTag: (tag: string) => string;
+  }
+
+  export interface ByteStream {
```
`ByteStream` does not look right. Where is `readByteStream` ?

malaterre replied:
and `readFixedString`, `readUint16`, `readUint32` and `seek`

malaterre commented on index.d.ts:
```diff
@@ -0,0 +1,108 @@
+    byteArray: ByteArray;
+    byteArrayParser: ByteArrayParser;
+    position: number;
+    warnings: string[];
```
Same comment for `warnings` as above

sedenardi replied:
The `warnings` array property is always present on a `ByteStream`.

malaterre commented:
@dannyrb This look like a pretty good stard. But the OP did not really describe the scope of this PR. For example a lof of the exported symbols are not describe in this PR:

* https://github.com/cornerstonejs/dicomParser/blob/374ad3d20a92bd2a7ec56d90dfce2fb2815fee13/src/index.js

Maybe this will be done at a later stage, but it would be nice to know the plan.

sedenardi commented:
@malaterre @dannyrb Thanks for the feedback guys. You're right that I didn't really define the scope of what I'd initially work on, mostly because copying (and modifying) the definitions from https://github.com/cornerstonejs/dicomParser/issues/124 satisfied my own project's needs.

Instead, I've completed the definitions for all the exported symbols. I haven't added annotations to them right now, but that could be done at a later date (mostly by copy/pasting the JSDoc).

Let me know if you guys see anything else that's off.

dannyrb commented:
I'll pull the lever, and if anything else jumps out we can address it.

sedenardi commented:
Thanks @dannyrb. I'll be happy to address any issues that arise from TS definitions in the future.

adamwdennis commented:
@sedenardi thank you for this -- any plans on making this more generally available via [DefinitelyTyped](https://definitelytyped.org/)?

sedenardi commented:
@adamwdennis Not sure what you mean by more generally available? I believe the DefinitelyTyped folks prefer the type definitions to only be in one place, the package or the DefinitelyTyped repo, and of those two options I think they prefer the package itself.

[From](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package)
> If you are the library author and your package is written in TypeScript, bundle the autogenerated declaration files in your package instead of publishing to Definitely Typed.

[From](https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package)
> When a package bundles its own types, types should be removed from Definitely Typed to avoid confusion.

sedenardi commented:
Hi @dannyrb, it doesn't look like CI automatically cut a new npm release. There's no huge rush, I just wanted to make sure nothing additional was needed on my end.
